snuba/query/matcher.py

```diff
@@ -0,0 +1,218 @@
+        return MatchResult()
+
+
+@dataclass(frozen=True)
+class String(Pattern[Optional[str]]):
+    """
+    Matches one specific string (or None).
+    """
+
+    value: Optional[str]
+
+    def match(self, node: Optional[str]) -> Optional[MatchResult]:
+        return MatchResult() if node == self.value else None
+
+
+@dataclass(frozen=True)
+class Or(Pattern[TNode], Generic[TNode]):
+    """
+    Union of multiple patterns. Matches if at least one is a valid match
+    and returns the first valid one.
+    """
+
+    patterns: Sequence[Pattern[TNode]]
+
+    def match(self, node: TNode) -> Optional[MatchResult]:
+        for p in self.patterns:
+            ret = p.match(node)
+            if ret:
+                return ret
+        return None
+
+
+@dataclass(frozen=True)
+class Column(Pattern[Expression]):
+    """
+    Matches a Column in an AST expression.
+    The column is defined by alias, name and table. For each a Pattern can be
+    provided. If one of these Patters is left None, that field will be ignored
+    when matching (equivalent to Any, but less verbose).
+    """
+
+    alias: Optional[Pattern[Optional[str]]] = None
+    column_name: Optional[Pattern[Optional[str]]] = None
+    table_name: Optional[Pattern[Optional[str]]] = None
+
+    def match(self, node: Expression) -> Optional[MatchResult]:
+        if not isinstance(node, ColumnExpr):
+            return None
+
+        result = MatchResult()
+        for pattern, value in (
+            (self.alias, node.alias),
+            (self.column_name, node.column_name),
+            (self.table_name, node.table_name),
+        ):
+            if pattern:
+                partial_result = pattern.match(value)
+                if not partial_result:
+                    return None
+                result = result.merge(partial_result)
+
+        return result
+
+
+@dataclass(frozen=True)
+class Function(Pattern[Expression]):
+    """
+    Matches a Function in the AST expression.
+    It works like the Column Pattern. if alias, function_name and function_parameters
+    are provided, they have to match, otherwise they are ignored.
+    """
+
+    alias: Optional[Pattern[Optional[str]]] = None
+    function_name: Optional[Pattern[Optional[str]]] = None
+    function_parameters: Optional[Tuple[Pattern[Expression], ...]] = None
+
+    def match(self, node: Expression) -> Optional[MatchResult]:
+        if not isinstance(node, FunctionCall):
+            return None
+
+        result = MatchResult()
+        for pattern, value in (
+            (self.alias, node.alias),
+            (self.function_name, node.function_name),
+        ):
+            if pattern:
+                partial_result = pattern.match(value)
+                if not partial_result:
+                    return None
+                result = result.merge(partial_result)
+
+        if self.function_parameters:
+            if len(self.function_parameters) != len(node.parameters):
+                return None
```

Comment:
This seems kind of dangerous when dealing with default arguments.

Reply:
Since FunctionCall does not have a concept of mandatory argument and optional argument (it represent a function call and not a function definition so I don't think that is wrong) I believe the only way we could deal with them is by defining a tuple of mandatory parameters that have to match and a tuple of optional parameters that would be ignored.
Though, in order to use this effectively, you'd always have to provide, when instantiating a Pattern all the optional parameters that function may take, which I guess most people would not do (it would be very easy to forget to add a bunch of Any() at the end of the expression).

What about this instead, just letting the user decide whether the list of parameters they want to match is `strict` or `not strict`. If it is strict, then all parameters must match. If it is not, all the parameters provided have to match, if there are further parameters in the FunctionCall they are ignored. What do you think ?

Reply:
> What about this instead, just letting the user decide whether the list of parameters they want to match is `strict` or `not strict`. If it is strict, then all parameters must match. If it is not, all the parameters provided have to match, if there are further parameters in the FunctionCall they are ignored. What do you think ?

This seems like a good idea to me. 👍